lib/fa2/nft/nft.impl.mligo

```diff
@@ -3,260 +3,73 @@
-        Some (opSet) -> opSet
-      | None -> Set.empty in
-    Set.size authorized > 0n || op.owner = op.operator
-
-  let token_metadata (type a) (p : nat) (s : a storage) : TZIP12.tokenMetadataData =
-    match Big_map.find_opt p s.token_metadata with
-      Some (data) -> data
-    | None () -> failwith Errors.undefined_token
-
-end
-
-module NFT = struct
-  type ledger = NFTExtendable.ledger
-
-  type operator = NFTExtendable.operator
-
-  type operators = NFTExtendable.operators
-
-  type storage = unit NFTExtendable.storage
-
-  type ret = unit NFTExtendable.ret
-
-  [@entry]
-  let transfer (t : TZIP12.transfer) (s : storage) : ret =
-    NFTExtendable.transfer t s
-
-  [@entry]
-  let balance_of (b : TZIP12.balance_of) (s : storage) : ret =
-    NFTExtendable.balance_of b s
-
-  [@entry]
-  let update_operators (updates : TZIP12.update_operators) (s : storage) : ret =
-    NFTExtendable.update_operators updates s
-
-  [@view]
-  let get_balance (p : (address * nat)) (s : storage) : nat =
-    NFTExtendable.get_balance p s
-
-  (* FIXME Not sure why we are implementing the two following views *)
-  [@view]
-  let total_supply (token_id : nat) (s : storage) : nat =
-    NFTExtendable.total_supply token_id s
-
-  [@view]
-  let all_tokens (_ : unit) (s : storage) : nat set =
-    NFTExtendable.all_tokens () s
-
-  [@view]
-  let is_operator (op : TZIP12.operator) (s : storage) : bool =
-    NFTExtendable.is_operator op s
-
-  [@view]
-  let token_metadata (p : nat) (s : storage) : TZIP12.tokenMetadataData =
-    NFTExtendable.token_metadata p s
-end
+#import "./extendable_nft.impl.mligo" "NFTExtendable"
+
+type ledger = NFTExtendable.ledger
+
+type operator = NFTExtendable.operator
+
+type operators = NFTExtendable.operators
+
+type storage = {
+   ledger : ledger;
+   operators : operators;
+   token_metadata : TZIP12.tokenMetadata;
+   metadata : TZIP16.metadata;
+}
+
+type ret = operation list * storage
+
+[@inline]
+let lift (s : storage) : unit NFTExtendable.storage =
+  {
+    ledger = s.ledger;
+    operators = s.operators;
+    token_metadata = s.token_metadata;
+    metadata = s.metadata;
+    extension = ();
+  }
+
+[@inline]
+let unlift (ret : operation list * unit NFTExtendable.storage) : ret =
+  let ops, s = ret in
+  ops,
+  {
+    ledger = s.ledger;
+    operators = s.operators;
+    token_metadata = s.token_metadata;
+    metadata = s.metadata;
+  }
+
+[@entry]
+let transfer (t : TZIP12.transfer) (s : storage) : ret =
+  unlift (NFTExtendable.transfer t (lift s))
+
+[@entry]
+let balance_of (b : TZIP12.balance_of) (s : storage) : ret =
+  unlift (NFTExtendable.balance_of b (lift s))
+
+[@entry]
+let update_operators (updates : TZIP12.update_operators) (s : storage) : ret =
+  unlift (NFTExtendable.update_operators updates (lift s))
+
+[@view]
+let get_balance (p : (address * nat)) (s : storage) : nat =
+  NFTExtendable.get_balance p (lift s)
+
+(* FIXME Not sure why we are implementing the two following views *)
```

Comment:
It's because we want to be compliant with https://tzip.tezosagora.org/proposal/tzip-12/
See 
```
Within its TZIP-016 metadata, an FA2 contract does not have to provide any off-chain-view but can provide 5 optional views:    

get_balance, total_supply, all_tokens, is_operator, and token_metadata.     

If present, all of these SHOULD be implemented, at least, as “Michelson Storage Views” and have the following types (Michelson annotations are optional) and semantics:
```

Reply:
Of course! Thank you.